src/protocols/secure_channel/CASESession.cpp

```diff
@@ -479,67 +524,82 @@ CHIP_ERROR CASESession::SendSigmaR2()
+        SuccessOrExit(err = tlvWriter.Finalize(&msg_R2_Signed));
     }
 
     // Step 7
-    err = mOpCredSet->SignMsg(mTrustedRootId, msg_R2_Signed.Get(), msg_r2_signed_len, sigmaR2Signature);
+    err = mOpCredSet->SignMsg(mTrustedRootId, msg_R2_Signed->Start(), msg_R2_Signed->DataLength(), sigmaR2Signature);
     SuccessOrExit(err);
 
     // Step 8
     msg_r2_signed_enc_len = sizeof(uint16_t) + mOpCredSet->GetDevOpCredLen(mTrustedRootId) + sigmaR2Signature.Length();
 
-    VerifyOrExit(msg_R2_Encrypted.Alloc(msg_r2_signed_enc_len), err = CHIP_ERROR_NO_MEMORY);
-    {
-        Encoding::LittleEndian::BufferWriter bbuf(msg_R2_Encrypted.Get(), msg_r2_signed_enc_len);
-
-        bbuf.Put16(mOpCredSet->GetDevOpCredLen(mTrustedRootId));
-        bbuf.Put(mOpCredSet->GetDevOpCred(mTrustedRootId), mOpCredSet->GetDevOpCredLen(mTrustedRootId));
-        bbuf.Put(sigmaR2Signature, sigmaR2Signature.Length());
+    msg_R2_Encrypted = System::PacketBufferHandle::New(msg_r2_signed_enc_len + System::PacketBuffer::kDefaultHeaderReserve);
+    VerifyOrExit(!msg_R2_Encrypted.IsNull(), err = CHIP_SYSTEM_ERROR_NO_MEMORY);
 
-        VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_NO_MEMORY);
+    // Generate Sigma2 TBE Data
+    {
+        System::PacketBufferTLVWriter tlvWriter;
+        TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
+
+        tlvWriter.Init(std::move(msg_R2_Encrypted));
+        SuccessOrExit(tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+        SuccessOrExit(err = tlvWriter.PutBytes(CASETLVTag::kNOC, mOpCredSet->GetDevOpCred(mTrustedRootId),
+                                               mOpCredSet->GetDevOpCredLen(mTrustedRootId)));
+        SuccessOrExit(
+            err = tlvWriter.PutBytes(CASETLVTag::kSignature, sigmaR2Signature, static_cast<uint32_t>(sigmaR2Signature.Length())));
+        SuccessOrExit(err = tlvWriter.EndContainer(outerContainerType));
+        SuccessOrExit(err = tlvWriter.Finalize(&msg_R2_Encrypted));
     }
 
     // Step 9
-    err = AES_CCM_encrypt(msg_R2_Encrypted.Get(), msg_r2_signed_enc_len, nullptr, 0, sr2k, kAEADKeySize, kIVSR2, kIVLength,
-                          msg_R2_Encrypted.Get(), tag, sizeof(tag));
+    err = AES_CCM_encrypt(msg_R2_Encrypted->Start(), msg_R2_Encrypted->DataLength(), nullptr, 0, sr2k, kAEADKeySize, kIVSR2,
+                          kIVLength, msg_R2_Encrypted->Start(), tag, sizeof(tag));
     SuccessOrExit(err);
 
     data_len = static_cast<uint16_t>(kSigmaParamRandomNumberSize + sizeof(uint16_t) + kTrustedRootIdSize + kP256_PublicKey_Length +
-                                     msg_r2_signed_enc_len + sizeof(tag));
+                                     msg_R2_Encrypted->DataLength() + sizeof(tag));
 
-    msg_R2 = System::PacketBufferHandle::New(data_len);
+    msg_R2 = System::PacketBufferHandle::New(data_len + System::PacketBuffer::kDefaultHeaderReserve);
```

Comment:
Isn't `System::PacketBuffer::kDefaultHeaderReserve` automatically added by default when New() is called?

Similar comment for other placed in this PR.

Reply:
Updated.